### ISSUE TITLE
roundabout exist counts are confusing for cyclists

### ISSUE THREAD

**Issue by emiltin** [Project-OSRM/osrm-backend#2533]
Because you can usually push bikes against oneways on bike, dual carriageways are counted as 2 exists in roundabouts, leading to confusing exit counts.

For example you're told to take the 3rd exit in the following situation:

![screen shot 2016-06-11 at 23 11 58](https://cloud.githubusercontent.com/assets/66034/15987483/1a98d0e4-302a-11e6-800f-abb3e26c6032.png)

I'm not sure what can be done about this. OSRM could count only exist that you can cycle on. This would make the common case of dual carriageways more correct, but would require special handling of edge cases like exiting on a footway.

**Comment by emiltin** [comment]
related to #2529. ie you could maybe tell the user "At roundabout, head straight on Birch Street" or "At roundabout, head left onto Summer Street",  instead of counting exits

**Comment by MoKob** [comment]
The problem I see here is how would you tell someone to exit into one of these streets. Consider a foot-path that is actually your destination. If you don't count it, you have either an undefined exit or, if you then simply add 1, `exit two` can both be the path where you need to push and the actual street.

What I could imagine is that we can treat roundabouts as we do with cars, not allowing exits against the direction, and have to announce something without an exit like `exit the roundabout and push your bike` for all exits that don't follow the convention.

This will require both special treatment in pre and post processing, since we need to handle cases that exit the roundabout at a exit that is not part of the roundabout and we need to consider travel modes of bikes in roundabouts.
This needs https://github.com/Project-OSRM/osrm-backend/issues/2113 to be implemented reasonably without introducing yet more profile dependent settings.

**Comment by emiltin** [comment]
the demo server seems to have a problem t the moment (no route is computed), but i was looking at these three roundabouts. there's clear main way that continues straight ahead in all of them.

http://map.project-osrm.org/?z=15&center=55.655110%2C12.587092&loc=55.659650%2C12.583508&loc=55.653016%2C12.588658&hl=en&alt=0

**Comment by emiltin** [comment]
Right, it's not obvious how to best guide cyclist in these situations. The question is how meaningful it is to count exits if you can exit pushing bikes against oneways or at footways.

For small roundabouts, simply announcing a normal turn is probably better, and is already done with the roundabout turn instruction.

For normal/large roundabouts, maybe it would be better to announce the exit as a normal turn? Ie. in 20m exit the roundabout onto ...?

I guess pedestrian routing have the same issue with roundabouts.

**Comment by Ajojo44** [comment]
Hi everyone,

I have the same problem with round-abouts. The interpretation of the exit is not good for our system as a instruction to give to cyclists. Most of the cyclists think as car drivers when they are on the road and as a car-driver you only count exits you could take on a round-about.

I understand that if the API returns a wrong exit based on this interpretation of the round-abouts, it's because all outputs are counted as user can ride down of its bike to walk on that way. 
This could be really dangerous on the road and lead to real security problems and therefore responsibility problems for us.

I see two or three workarounds for this problem:

- First solution could be to propose two kinds of cyclist profiles: CYCLING ONLY and CYCLING/WALKING. With that separation of the profiles, we will still have the same behavior with the CYCLING/WALKING profile but we will have a filter on the CYCLING ONLY profile that will not take pedestrian-only ways as accessible. Exits on a round-about will therefore be only real available exits as a cyclist.

- Another workaround would be to add a new field on each round-about exit that would precise which is the profile concerned ("driving", "cycling" or "walking"). Like that, everyone would be able to make its own interpretation of the exit number. But ... this also means that the exit could be a WALKING exit and that we only counted NON-WALKING exits. Therefore, a special message should be given to the user in order to inform him that on this particular round-about, he should ride down of his bicycle to walk to a particular exit and maybe have a look on the map as its a different behaviour.

- If you don't want to add a new profile such as CYCLING-ONLY, you can maybe add an optional parameter on the request "roundabout_profile_exit_type" that would avoid the API to guide the user to certain exit types. Example: "roundabout_profile_exit_type={driving, cycling, walking}" or to solve our current problem: "roundabout_profile_exit_type={driving, cycling}" (walking exits are filtered). One of the main advantage of this solution is to keep simple turn instructions to a track available as its less dangerous and easier to indicate to the user.

Do you think one of these solutions is conceivable? How long would it take before seeing a satisfactory solution? For us it's a real issue that leads us to a deadlock.

Thank you!

**Comment by danpat** [comment]
One low-hanging fruit that might be an option would be to not count exits that would trigger a mode change.

For the car, this would mean nothing would change - unless there's a roundabout directly connected to a ferry somewhere.

For bike, we could suppress counting of any exits with a mode that differs from the roundabout mode.

For walk, I'm not sure what the right thing do do is - do walkers perceive the car exit numbers the same?  Personally, I would count "exits" as if I was driving - roundabouts are car constructs, after all.

What I'm not sure how to handle with this solution is what exit number to assign a route where we *do* suggest a mode change - "take exit 3.5?"

**Comment by MoKob** [comment]
@danpat there is at least one of these roundabouts into ferries: https://github.com/Project-OSRM/osrm-backend/pull/3766

**Comment by emiltin** [comment]
maybe osrm could announce "exit on the footpath after the 2nd exit?"

**Comment by mattfbacon** [comment]
We just encountered the exit counting issue with bicycles. Here was the discussion on OSM (In german): <https://www.openstreetmap.org/note/4470785#map=19/49.902639/10.067153&layers=N>. Has anything been done to address this? I don't understand why it would be different for bicycles because they also have to follow traffic rules like one-way streets. And no, you can't get off your bike in the middle of a roundabout and push it the wrong way on the roundabout entrance road.
